Make BackgroundService run ExecuteAsync as task

ericstj opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/36063

That issue requested that we move the entire `BackgroundService.ExecuteAsync` to the async part of the method rather than give the service a chance to do synchronous work before it's first await.  In this way `BackgroundService`s will not hold up the start of the application and folks are free to implement ExecuteAsync as a completely blocking method.

This is a breaking change, but one that was deemed acceptable.

ericstj committed:
Make BackgroundService run ExecuteAsync as task

ericstj commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/BackgroundServiceTests.cs:
```diff
@@ -26,28 +26,12 @@ public void StartReturnsCompletedTaskIfLongRunningTaskIsIncomplete()
-        public void StartReturnsCompletedTaskIfCancelled()
+        public async Task StartCancelledThrowsTaskCanceledException()
         {
-            var tcs = new TaskCompletionSource<object>();
-            tcs.TrySetCanceled();
-            var service = new MyBackgroundService(tcs.Task);
-
-            var task = service.StartAsync(CancellationToken.None);
-
-            Assert.True(task.IsCompleted);
-            Assert.Same(task, service.ExecuteTask);
-        }
-
-        [Fact]
-        public async Task StartReturnsLongRunningTaskIfFailed()
-        {
-            var tcs = new TaskCompletionSource<object>();
-            tcs.TrySetException(new Exception("fail!"));
-            var service = new MyBackgroundService(tcs.Task);
-
-            var exception = await Assert.ThrowsAsync<Exception>(() => service.StartAsync(CancellationToken.None));
```
These two tests are no longer applicable since none of ExecuteAsync is executed as part of the StartAsync.

ericstj commented:
> Should we expose a config workaround for this breaking change

That was specifically covered in https://github.com/dotnet/runtime/issues/36063 and the API proposal around it.  The API to control this was rejected and as a rule we don't proactively add AppContext compat switches until we have a strong signal that we need them.  The belief here is that most folks will prefer that their `BackgroundTask` has the entirety of its `ExecuteAsync` method run on the background thread.  Folks already have the workaround to override `StartAsync` and do some work before calling `base.StartAsync`, implement [IHostedLifecycleService ](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedlifecycleservice?view=net-9.0-pp) on their BackGroundService, or forgo BackgroundService and implement their own [IHostedService](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice?view=net-9.0-pp)

ericstj committed:
Simplify BackGroundService.StartAsync, fix tests

ericstj commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs:
```diff
@@ -688,7 +688,7 @@ public async Task WebHostStopAsyncUsesDefaultTimeoutIfNoTokenProvided()
 
         [Fact]
-        public async Task HostPropagatesExceptionsThrownWithBackgroundServiceExceptionBehaviorOfStopHost()
+        public async Task HostStopsWithBackgroundServiceExceptionBehaviorOfStopHost()
```
So this wasn't guaranteed before, it was only happening because the test passed `Task.CompletedTask` so the exception happened synchronously in the call to `ExecuteAsync` - no propagation was happening, it was just normal exception behavior.

There's nothing here that actually is meant to propagate an exception to a different thread, just stop the service:https://github.com/dotnet/runtime/blob/05e5a448fe726617ed881620093a792f27371e71/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L187-L204

So I've adjusted this test case to just ensure that the lifetime stopping is triggered.

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/BackgroundServiceTests.cs:
```diff
@@ -26,28 +26,12 @@ public void StartReturnsCompletedTaskIfLongRunningTaskIsIncomplete()
+            var service = new WaitForCancelledTokenService();
 
-            Assert.Equal("fail!", exception.Message);
+            await Assert.ThrowsAsync<TaskCanceledException>(() => service.StartAsync(ct));
```
```suggestion
            await Assert.ThrowsAnyAsync<OperationCanceledException>(() => service.StartAsync(ct));
```

ericstj committed:
Merge branch 'main' of https://github.com/dotnet/runtime into backgroundServiceFullStartAsync

ericstj committed:
Never return ExecuteAsync task from StartAsync

It's possible the ExecuteAsync task completes in the time between it was
started and StartAsync returns.  If so, we were returning it's task.
This leads to non-deterministic error behavior of the StartAsync API as
well as inconsistently honoring `BackgroundServiceExceptionBehavior`
based on this race.

Instead always return a completed task.

ericstj committed:
Address feedback

ericstj commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs:
```diff
@@ -42,8 +42,8 @@ public virtual Task StartAsync(CancellationToken cancellationToken)
+            _executeTask = Task.Run(() => ExecuteAsync(_stoppingCts.Token), _stoppingCts.Token);
 
             // If the task is completed then return it, this will bubble cancellation and failure to the caller
             if (_executeTask.IsCompleted)
```
@stephentoub let me know if you think this is acceptable.  It seems to me like it's consistent with the goal of the issue to move all the work to happen asynchronously / in the background.

If there's a better way to distinguish a synchronous failure to create the task and surface that we could, but the previous solution of returning the task from Task.Run was causing an ExecuteAsync exception to happen synchronously based on the timing of the background Task.

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs:
```diff
@@ -39,19 +39,15 @@ public abstract class BackgroundService : IHostedService, IDisposable
         /// <returns>A <see cref="Task"/> that represents the asynchronous Start operation.</returns>
         public virtual Task StartAsync(CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
```
What's the reason for adding this? If we do want this check, it should instead be:
```C#
if (cancellationToken.IsCancellationRequested)
{
    return Task.FromCanceled(cancellationToken);
}
```

ericstj replied:
I wanted to preserve the behavior of throwing synchronously when calling the method with an already cancelled token.

> if we do want this check, it should instead be

Interesting, the reason I thought `ThrowIfCancellationRequested` was acceptable was because you previously said this https://github.com/dotnet/runtime/pull/116283#discussion_r2131453586.  I agree though that I need Task.FromCanceled if I want the same exception as before.  I'll adjust it.

stephentoub replied:
My comment isn't about the type of the exception but rather whether the exception propagates out of the method synchronously or asynchronously. We generally consider it a bug if non-argument exceptions propagate out of a Task-returning method synchronously.

ericstj replied:
I guess this is in a similar boat as above - there's a race here where we check and throw - however that I'm willing to accept that race since it's necessary to capture the case where there was no race at all (already canceled token).

ericstj committed:
Address feedback

ericstj committed:
Expose cancellation async

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs:
```diff
@@ -42,16 +42,10 @@ public virtual Task StartAsync(CancellationToken cancellationToken)
 
-            // Store the task we're executing
-            _executeTask = ExecuteAsync(_stoppingCts.Token);
+            // Execute all of ExecuteAsync as a background thread, and store the task we're executing so that we can wait for it later.
```
```suggestion
            // Execute all of ExecuteAsync asynchronously, and store the task we're executing so that we can wait for it later.
```

stephentoub commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs:
```diff
@@ -702,7 +702,12 @@ public async Task HostPropagatesExceptionsThrownWithBackgroundServiceExceptionBe
+            await host.StartAsync();
+
+            // host is expected to catch exception, then trigger ApplicationStopping.
+            // give the host 1 minute to stop.
```
I assume it's usually way way faster than that?

ericstj committed:
Update src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>
